pkg/agent/plugin/svidstore/gcloudsecretmanager/gcloud.go

```diff
@@ -0,0 +1,246 @@
+	secret, err := getSecret(ctx, client, opt.secretName())
+	if err != nil {
+		return nil, err
+	}
+
+	// Secret not found, create it
+	if secret == nil {
+		secret, err = client.CreateSecret(ctx, &secretmanagerpb.CreateSecretRequest{
+			Parent:   opt.parent(),
+			SecretId: opt.name,
+			Secret: &secretmanagerpb.Secret{
+				// TODO: what replication type must we use here?
+				Replication: &secretmanagerpb.Replication{
+					Replication: &secretmanagerpb.Replication_Automatic_{
+						Automatic: &secretmanagerpb.Replication_Automatic{},
+					},
+				},
+				Labels: map[string]string{
+					"spire-svid": "true",
+				},
+			},
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create secret: %v", err)
+		}
+		p.log.With("secret_name", secret.Name).Debug("Secret created")
+	}
+
+	secretData, err := svidstore.SecretFromProto(req)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to parse request: %v", err)
+	}
+
+	secretBinary, err := json.Marshal(secretData)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to parse payload: %v", err)
+	}
+
+	resp, err := client.AddSecretVersion(ctx, &secretmanagerpb.AddSecretVersionRequest{
+		Parent: secret.Name,
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: secretBinary,
+		},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to add secret version: %v", err)
+	}
+
+	p.log.With("state", resp.State).With("name", resp.Name).Debug("Secret payload updated")
+
+	return &svidstorev1.PutX509SVIDResponse{}, nil
+}
+
+// DeleteX509SVID deletes a Secret in the configured Google Cloud Secrets manager
+func (p *SecretManagerPlugin) DeleteX509SVID(ctx context.Context, req *svidstorev1.DeleteX509SVIDRequest) (*svidstorev1.DeleteX509SVIDResponse, error) {
+	opt, err := optionsFromSecretData(req.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := p.hooks.newClient(ctx, p.config.ServiceAccountFile)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create secretmanager client: %v", err)
+	}
+	defer client.Close()
+
+	secret, err := getSecret(ctx, client, opt.secretName())
+	if err != nil {
+		return nil, err
+	}
+
+	if secret == nil {
+		p.log.With("secret_name", opt.secretName()).Warn("Secret to delete not found")
+		return &svidstorev1.DeleteX509SVIDResponse{}, nil
+	}
+
+	if err := client.DeleteSecret(ctx, &secretmanagerpb.DeleteSecretRequest{
+		Name: secret.Name,
+		Etag: secret.Etag,
+	}); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to delete secret: %v", err)
+	}
+
+	p.log.With("secret_name", opt.secretName()).Debug("Secret deleted")
+	return &svidstorev1.DeleteX509SVIDResponse{}, nil
+}
+
+// getSecret gets secret from GCloud and validtes if it has `spire-svid` label, nil if not found
+func getSecret(ctx context.Context, client secretsClient, secretName string) (*secretmanagerpb.Secret, error) {
```

Comment:
> another option is to add a selector that allow us to create a new tag with some information we must validate before update/delete a secret. what do you think about that approach?

I was wondering if we should set trust domain name or something like that as part of the tag value we already use

> secrets manager api didnt allow us to assign a policy on creation 'but' we can use SetIamPolicy, that will requiere some more information on selectors

Yes I think we will need something like this. The policy is a big JSON blob though, it would be much better if we could reference some pre-existing policy like the way the AWS plugin references a pre-existing key id

Reply:
sounds good to me, so 2 changes:

- spire-svid = td.IDString()
-  add new selector that allow us to get iam policy (optional?) and call SetIAMPolicy when we create/update an SVID.